### PR TITLE
T/396 TreeModel serialization / deseralization.

### PR DESCRIPTION
**Pull request by jodator**
This PR introduces converting TreeModel deltas to JSON and from JSON objects defined in #396.

### CODE REVIEW

File: src/treemodel/delta/delta.js [hunk 1]
```diff
@@ -118,6 +131,31 @@ export default class Delta {
...
+	 * Custom toJSON method to make deltas serializable.
+	 *
+	 * @returns {Object} Clone of this delta with added class name.
+	 */
+	toJSON() {
+		let json = clone( this );
+
+		json.__class = this.constructor.className;
+
+		json.batch = null;
```
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: `null` will end up in the JSON... shouldn't we delete this prop instead?

File: src/treemodel/delta/delta.js [hunk 2]
```diff
@@ -118,6 +131,31 @@ export default class Delta {
...
 
 	/**
+	 * Custom toJSON method to make deltas serializable.
+	 *
+	 * @returns {Object} Clone of this delta with added class name.
+	 */
+	toJSON() {
+		let json = clone( this );
+
+		json.__class = this.constructor.className;
```
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: Can't we call it `__className` for consistency?

File: src/treemodel/delta/delta.js [hunk 3]
```diff
@@ -118,6 +131,31 @@ export default class Delta {
...
+		json.batch = null;
+
+		return json;
+	}
+
+	/**
+	 * Delta class name. Used by {@link engine.treeModel.delta.Delta#toJSON} method for serialization and
+	 * {@link engine.treeModel.delta.Delta.fromJSON} during deserialization.
+	 *
+	 * @type {String}
```
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: Lacks `@readonly`.

File: src/treemodel/delta/delta.js (hunk 1 above)
Comment (jodator) [inline src/treemodel/delta/delta.js]: This can be changed, but should be changed in other places like operation where we have:
```js
json.parent = this.parent ? this.parent.name : null;
```
So this is consistent with others (setting null). So above should be changed as:

```js
if( this.parent ) {
    json.parent = this.parent.name;
} else {
    delete json.parent;
}
```

File: src/treemodel/delta/delta.js (hunk 2 above)
Comment (jodator) [inline src/treemodel/delta/delta.js]: I've used `__class` to be shorter but this is not so critical so can be `__className` also.

File: src/treemodel/delta/delta.js [hunk 4]
```diff
@@ -128,4 +166,64 @@ export default class Delta {
 	static get _priority() {
 		return 0;
 	}
+
+	/**
+	 * Creates InsertDelta from deserialized object, ie. from parsed JSON string
```
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: s/ie./i.e./

Comment (Reinmar) [inline src/treemodel/delta/delta.js]: s/$/.$/

File: src/treemodel/delta/delta.js (hunk 3 above)
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: May lack `@static` (not sure whether JSDoc is able to read automatically that property is static).

File: src/treemodel/delta/delta.js [hunk 5]
```diff
@@ -128,4 +166,64 @@ export default class Delta {
...
+operations[ MoveOperation.className ] = MoveOperation;
+operations[ NoOperation.className ] = NoOperation;
+operations[ Operation.className ] = Operation;
+operations[ ReinsertOperation.className ] = ReinsertOperation;
+operations[ RemoveOperation.className ] = RemoveOperation;
+operations[ RootAttributeOperation.className ] = RootAttributeOperation;
+
+const deserializers = new Map();
+
+export function registerDeserializer( className, constructor ) {
```
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: API docs are missing.

File: src/treemodel/delta/delta.js [hunk 6]
```diff
@@ -128,4 +166,64 @@ export default class Delta {
...
+	if ( deserializers.has( className ) ) {
```
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: Is it really needed? Deserializers will not be frequently added, so there's a very little chance for a bug.

File: src/treemodel/delta/insertdelta.js [hunk 7]
```diff
@@ -8,6 +8,7 @@
 import Delta from './delta.js';
 import RemoveDelta from './removedelta.js';
 import { register } from '../batch.js';
+import { registerDeserializer } from './delta.js';
```
Comment (Reinmar) [inline src/treemodel/delta/insertdelta.js]: You can merge that with the first import.

File: src/treemodel/delta/mergedelta.js [hunk 8]
```diff
@@ -47,6 +48,10 @@ export default class MergeDelta extends Delta {
 	get _reverseDeltaClass() {
 		return SplitDelta;
 	}
+
+	static get className() {
```
Comment (Reinmar) [inline src/treemodel/delta/mergedelta.js]: As proposed recently, let's document all inherited props and methods with `/** @inheritDoc */`.

File: src/treemodel/delta/delta.js (hunk 6 above)
Comment (jodator) [inline src/treemodel/delta/delta.js]: Not sure which part.
Every delta needs deserializer (even custom made by other devs). Also this part mirrors behavior from `batch#register`.

Comment (jodator) [inline src/treemodel/delta/delta.js]: The operations part is to not use if-else block.

Comment (Reinmar) [inline src/treemodel/delta/delta.js]: I meant throwing error and validating anything here. It's not going to be an API which is used by hundreds thousand developers. There's very little chance that this API is actually used by anyone except us in (and few times only). Hence, validation, while being correct, makes no practical sense.

File: src/treemodel/node.js [hunk 9]
```diff
@@ -188,6 +188,9 @@ export default class Node {
 		// Due to circular references we need to remove parent reference.
 		json.parent = this.parent ? this.parent.name : null;
 
+		// Serialize attributes as Map object is represented as "{}" when parsing to JSON.
+		json._attrs = [ ...json._attrs ];
```
Comment (Reinmar) [inline src/treemodel/node.js]: First of all, `_attrs` is an invalid name, because we should not use such shorthand for a property name (there's a ticket to fix this across the API). Second, it should not be prefixed with underscore, because it's not private in that JSON (cause it's not an API, just data).

tl;dr – rename it to `attributes`.

File: src/treemodel/nodelist.js [hunk 10]
```diff
@@ -50,14 +51,18 @@ class NodeListText extends Text {
...
+	 * Custom toJSON method to solve child-parent circular dependencies when serializing NodeListText to JSON string.
 	 *
 	 * @returns {Object} Clone of this object with the parent property replaced with its name.
 	 */
 	toJSON() {
 		const json = clone( this );
 
-		json.parent = json.parent ? this.parent.name : null;
+		// Due to circular references we need to remove parent reference.
+		json.parent = this.parent ? this.parent.name : null;
```
Comment (Reinmar) [inline src/treemodel/nodelist.js]: Do we need parent at all? I guess @scofalik has been using it for debugging. But now we have different tools, so it's unnecessary here.

File: src/treemodel/nodelist.js [hunk 11]
```diff
@@ -293,7 +298,7 @@ export default class NodeList {
 
 		this._indexMap.splice( index, number );
 
-		for ( let i = index; i < this._indexMap.length ; i++ ) {
+		for ( let i = index; i < this._indexMap.length; i++ ) {
```
Comment (Reinmar) [inline src/treemodel/nodelist.js]: This change is not part of this ticket. Please report a separate ticket for it.






trollolololo ;P

File: src/treemodel/operation/operation.js [hunk 12]
```diff
@@ -81,11 +81,36 @@ export default class Operation {
...
+	/**
+	 * Name of the operation class used for serialization.
+	 *
+	 * @type {String}
+	 */
+	static get className() {
+		return 'engine.treeModel.operation.Operation';
+	}
+
+	/*jshint unused: true*/
```
Comment (Reinmar) [inline src/treemodel/operation/operation.js]: ?

File: src/treemodel/operation/operation.js [hunk 13]
```diff
@@ -81,11 +81,36 @@ export default class Operation {
...
 	 */
 	toJSON() {
-		const json = clone( this );
+		const json = clone( this, true );
+
+		json.__class = this.constructor.className;
 
 		// Due to circular references we need to remove parent reference.
-		json.delta = this.delta ? '[engine.treeModel.Delta]' : null;
+		json.delta = this.delta ? `[${this.delta.constructor.className}]` : null;
```
Comment (Reinmar) [inline src/treemodel/operation/operation.js]: Just like in https://github.com/ckeditor/ckeditor5-engine/pull/418/files#r63502600

File: src/treemodel/position.js [hunk 14]
```diff
@@ -564,6 +564,34 @@ export default class Position {
 	}
 
 	/**
+	 * Creates Element object from deserilized object, ie. from parsed JSON string.
+	 *
+	 * @param {Object} json Deserialized JSON object.
+	 * @param {engine.treeModel.Document} doc Document on which this operation will be applied.
```
Comment (Reinmar) [inline src/treemodel/position.js]: Should be `document`.

File: src/treemodel/delta/delta.js (hunk 6 above)
Comment (jodator) [inline src/treemodel/delta/delta.js]: I understand this, but how this compare to: https://github.com/ckeditor/ckeditor5-engine/blob/master/src/treemodel/batch.js#L118-L132 (which is in my understanding has the same frequency of usage as you described).

File: src/treemodel/nodelist.js (hunk 10 above)
Comment (jodator) [inline src/treemodel/nodelist.js]: If this is not needed anymore I could drop this across the ticket as in delta case.

File: src/treemodel/rootelement.js [hunk 15]
```diff
@@ -19,16 +19,35 @@ export default class RootElement extends Element {
 	 *
 	 * @param {engine.treeModel.Document} doc {@link engine.treeModel.Document} that is an owner of the root.
 	 * @param {String} name Node name.
+	 * @param {String} rootName Root name inside parent {@link engine.treeModel.Document}.
 	 */
-	constructor( doc, name ) {
+	constructor( doc, name, rootName ) {
```
Comment (Reinmar) [inline src/treemodel/rootelement.js]: @pjasiun, what do you think about making this constructor protected so only the document is able to create it (then we can be sure that names are synchronised).

File: tests/treemodel/node.js [hunk 16]
```diff
@@ -178,4 +179,58 @@ describe( 'Node', () => {
 			} );
 		} );
 	} );
+
+	describe( 'toJSON', () => {
+		it( 'should serialize empty node', () => {
+			let node = new Element( 'one' );
```
Comment (Reinmar) [inline tests/treemodel/node.js]: Why do you test serialising element in node's tests?

File: tests/treemodel/node.js [hunk 17]
```diff
@@ -178,4 +179,58 @@ describe( 'Node', () => {
...
+						{ _attrs: [], _children: { _indexMap: [], _nodes: [] }, name: 'one', parent: null },
+						{
+							_attrs: [], _children: {
+							_indexMap: [ 0, 0, 1, 2 ], _nodes: [
+								{ _attrs: [], text: 'ba', parent: 'two' },
+								{ _attrs: [], _children: { _indexMap: [], _nodes: [] }, name: 'img', parent: 'two' },
+								{ _attrs: [], text: 'r', parent: 'two' }
+							]
+						}, name: 'two', parent: null
+						},
```
Comment (Reinmar) [inline tests/treemodel/node.js]: This object isn't correctly formatted. Two `}` at the same level of indentation mean that something is wrong here.

File: tests/treemodel/_utils/utils.js [hunk 18]
```diff
@@ -34,6 +34,10 @@ const utils = {
 		}
 
 		return txt;
+	},
+
+	jsonParseStringify( object ) {
```
Comment (Reinmar) [inline tests/treemodel/_utils/utils.js]: Lacks docs.

File: src/treemodel/rootelement.js [hunk 19]
```diff
@@ -19,16 +19,35 @@ export default class RootElement extends Element {
...
+		this.rootName = rootName;
+	}
+
+	/**
+	 * Custom toJSON method to solve child-parent circular dependencies.
+	 *
+	 * @method engine.treeModel.RootElement#toJSON
+	 * @returns {String} Name of this root inside {@link engine.treeModel.Document} that is an owner of this root.
+	 */
+	toJSON() {
```
Comment (Reinmar) [inline src/treemodel/rootelement.js]: Implementation of this method is confusing. Why is the name of the root enough in `toJSON()` to represent the root? Why there are no children or attributes of it (in fact – attributes are really missing)?

If some of these things are not necessary, this should be documented why. The fact that this implementation solve child-parent circ deps is a secondary aspect, and I'm not even sure whether it's true (because the same would need to be done for all elements, while in those cases it's handled by deleting `child.parent`).

**Comment by Reinmar** [comment]
I left you couple of comments here and there. Some things need to be cleaned up or adjusted a bit. Some may need clarifications. I think that the general topic which isn't defined anywhere is what needs to be stringified to represent the document in JSON. Perhaps you could write somewhere (in some most-outer stringifier or parser) few words about that.

**Comment by jodator** [comment]
Well there is no most-outer stringifier besides `JSON.parse( [ /* deltas */ ] );`.

Most-outer parser would be `Delta.fromJSON()`

File: src/treemodel/delta/delta.js (hunk 6 above)
Comment (Reinmar) [inline src/treemodel/delta/delta.js]: IMO that `register()` makes unnecessary validation as well. I'd be ok with removing it.

**Comment by Reinmar** [comment]
> Most-outer parser would be Delta.fromJSON()

Maybe it'd make sense to introduce `Serializer` class? It'd contain the whole documentation plus two `fromJSON` and `toJSON` methods which would basically execute related methods on passed items. I know that so far it's very little responsibility, but for some reason it looks like a cleaner solution anyway. WDYT?

**Comment by jodator** [comment]
Yep it might be as you described.

File: src/treemodel/delta/delta.js [hunk 20]
```diff
@@ -128,4 +166,64 @@ export default class Delta {
...
+				{ name: json.__className }
+			);
+		}
+
+		let Constructor = deserializers.get( json.__className );
+
+		let delta = new Constructor();
+
+		if ( json.operations.length ) {
+			json.operations.forEach( ( operation ) => delta.addOperation( operations[ operation.__className ].fromJSON( operation, doc ) ) );
```
Comment (pjasiun) [inline src/treemodel/delta/delta.js]: Use for-of instead of `forEach`. We try to avid `forEach` and use it only when it can not be simple replaced with for-of.

File: src/treemodel/delta/delta.js (hunk 3 above)
Comment (jodator) [inline src/treemodel/delta/delta.js]: `@static` is not needed.
selection_020

File: src/treemodel/delta/delta.js (hunk 6 above)
Comment (jodator) [inline src/treemodel/delta/delta.js]: If so then ok.

File: src/treemodel/delta/delta.js [hunk 21]
```diff
@@ -128,4 +167,64 @@ export default class Delta {
...
+}
+
+const operations = {};
+operations[ AttributeOperation.className ] = AttributeOperation;
+operations[ InsertOperation.className ] = InsertOperation;
+operations[ MoveOperation.className ] = MoveOperation;
+operations[ NoOperation.className ] = NoOperation;
+operations[ Operation.className ] = Operation;
+operations[ ReinsertOperation.className ] = ReinsertOperation;
+operations[ RemoveOperation.className ] = RemoveOperation;
```
Comment (pjasiun) [inline src/treemodel/delta/delta.js]: This is definitely not `Deltas` responsibility to handle all operations instanced. If there is no better place for it you should create `OperationFactory`.

Comment (pjasiun) [inline src/treemodel/delta/delta.js]: Or `OperationDeserializer`.

File: src/treemodel/node.js (hunk 9 above)
Comment (jodator) [inline src/treemodel/node.js]: It is not a part of this ticket I think. While I can change this to:

```js
json._attributes = [ ...json._attrs ];
```
I also need to remove:
```js
delete json._attrs;
```

tl;dr - this name mirrors object's properties.

File: src/treemodel/operation/operation.js (hunk 12 above)
Comment (jodator) [inline src/treemodel/operation/operation.js]: Base class doesn't need document - but all operations need document parameter.

Without `jshint unused: true` linter will throw errors.

This is to make methods consistent in documentation so all (base and child classes) heve the same signature.

File: src/treemodel/node.js (hunk 9 above)
Comment (pjasiun) [inline src/treemodel/node.js]: I think it is the bigger question: should serialised object keep private properties? I am not sure about the answers. It is simpler to use private properties, and this is what we do, for instance in `clone` or `isEqual` methods. On the other hand, serialised object, send over the network seems to be a public API. If it will use private properties then send objects will change every time private properties change, what looks like a bad smell.

File: src/treemodel/operation/operation.js (hunk 12 above)
Comment (pjasiun) [inline src/treemodel/operation/operation.js]: It is not possible that linter throw error on the public method.

File: src/treemodel/rootelement.js (hunk 15 above)
Comment (pjasiun) [inline src/treemodel/rootelement.js]: I do not find it necessarily. If someone creates root in the document using `createRoot`, this method will create `RootElement` with the same name. There is no other way to create root in the `Document`, so there is not way to desynchronise these names. Also If someone want to create a root element, for instance controller, temporarily, for some reason, then, why should we prevent it?

File: src/treemodel/rootelement.js [hunk 22]
```diff
@@ -19,16 +19,35 @@ export default class RootElement extends Element {
...
+		return typeof this.rootName === 'symbol' ? '$$graveyard' : this.rootName;
```
Comment (pjasiun) [inline src/treemodel/rootelement.js]: Why you assume that every symbol is a `graveyard`? You could replace it with `this === this.document.graveyard`. But  I think that we should simplified it in general and use `$graveyard` string instead of using symbols. Also one `$` should be enough.

File: src/treemodel/nodelist.js (hunk 10 above)
Comment (pjasiun) [inline src/treemodel/nodelist.js]: This code exists only because of objects thrown in errors, which can not have circular dependencies. The `parent` property can be removed.

File: tests/treemodel/node.js [hunk 23]
```diff
@@ -178,4 +179,58 @@ describe( 'Node', () => {
...
+				name: 'one',
+				parent: null
+			} );
+		} );
+
+		it( 'should serialize node with children', () => {
+			expect( treeModelTestUtils.jsonParseStringify( root ) ).to.deep.equal( {
+				_attrs: [],
+				_children: {
+					_indexMap: [ 0, 1, 2 ],
```
Comment (pjasiun) [inline tests/treemodel/node.js]: This is again question about private properties in serialised object. `_indexMap` is the private and internal mechanism to improve performance. If we change it none tests should fail. Now we would need to change these tests every time we change private class logics.

File: tests/treemodel/_utils/utils.js (hunk 18 above)
Comment (pjasiun) [inline tests/treemodel/_utils/utils.js]: Also it should return two separate exports instead of the `utils` object.

File: src/treemodel/delta/delta.js (hunk 21 above)
Comment (pjasiun) [inline src/treemodel/delta/delta.js]: But we where thinking a lot about `OperationFactory`, which might be useful also to resolve circular dependencies between operations, so I think it might be useful for more then deserialisation`.

File: src/treemodel/delta/delta.js (hunk 5 above)
Comment (pjasiun) [inline src/treemodel/delta/delta.js]: Similarly this function should be moved to the `DeltaFactory`.

**Comment by pjasiun** [comment]
To sum up, I think there are two most important changes:

 - making serialised object "public", so they should not keep private properties, anything what is not data. 
 - introducing `DeltaFactory` and `OperationFactory`.

File: src/treemodel/node.js (hunk 9 above)
Comment (jodator) [inline src/treemodel/node.js]: That's some point but we have this in more places:
```js
let node = new Element( 'one' );

expect( treeModelTestUtils.jsonParseStringify( node ) ).to.deep.equal( {
	_attrs: [],
	_children: {
		_indexMap: [],
		_nodes: []
	},
	name: 'one'
} );
```

So for consistency I've left `_attrs` but the question about all private properties remains. It would be nice of you (@reinmar, @pjasiun) to give me your definitive view on this: change them or not?

Comment (pjasiun) [inline src/treemodel/node.js]: Change. :)

File: src/treemodel/operation/operation.js (hunk 12 above)
Comment (jodator) [inline src/treemodel/operation/operation.js]: The linter throws error on unused parameter `document`

Comment (pjasiun) [inline src/treemodel/operation/operation.js]: So can't you remove this parameter here since it is unused?

**Comment by jodator** [comment]
All reported issues should be fixed now.

File: src/treemodel/delta/attributedelta.js [hunk 24]
```diff
@@ -69,6 +70,11 @@ export default class AttributeDelta extends Delta {
 		return AttributeDelta;
 	}
 
+	/** @inheritDoc */
```
Comment (pjasiun) [inline src/treemodel/delta/attributedelta.js]: For consistency it should be:
```
/**
 * @inheritDoc
 */
```

File: src/treemodel/delta/attributedelta.js [hunk 25]
```diff
@@ -82,7 +88,12 @@ export default class AttributeDelta extends Delta {
  * @memberOf engine.treeModel.delta
  * @extends engine.treeModel.delta.Delta
  */
-export class RootAttributeDelta extends Delta {}
+export class RootAttributeDelta extends Delta {
+	/** @inheritDoc */
```
Comment (pjasiun) [inline src/treemodel/delta/attributedelta.js]: Same here.

File: src/treemodel/delta/deltafactory.js [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+		}
+
+		return delta;
+	}
+}
+
+const deserializers = new Map();
+
+export function registerDeserializer( className, constructor ) {
+	deserializers.set( className, constructor );
```
Comment (pjasiun) [inline src/treemodel/delta/deltafactory.js]: Can we move this function to the `DeltaFactory` as a static method? Also it could take only one parameter `constructor`, class name can be taken from it. And I think that it could be generic deltas register, so `registerDeserializer` could be changed to simple `register`. I can imagine more usage then only deserialisation.

tl;dr: change:
```
registerDeserializer( Delta.className, Delta );
```
To:
```
DeltaFactory.register( Delta );
```

File: tests/treemodel/_utils/utils.js [hunk 27]
```diff
@@ -7,34 +7,39 @@
...
+ * Returns tree structure as a simplified string. Elements are uppercase and characters are lowercase.
+ * Start and end of an element is marked the same way, by the element's name (in uppercase).
+ *
+ *		let element = new Element( 'div', [], [ 'abc', new Element( 'p', [], 'foo' ), 'xyz' ] );
+ *		treemodelUtils.getNodesAndText( element ); // abcPfooPxyz
+ *
+ * @param {engine.treeModel.Range} range Range to stringify.
+ * @returns {String} String representing element inner structure.
+ */
+export default function getNodesAndText( range ) {
```
Comment (pjasiun) [inline tests/treemodel/_utils/utils.js]: Why this export is default? Is it more important for some reasons?

File: tests/treemodel/element.js [hunk 28]
```diff
@@ -233,10 +234,102 @@ describe( 'Element', () => {
...
+
+			let serialized = jsonParseStringify( el );
+
+			let deserialized = Element.fromJSON( serialized );
+
+			expect( deserialized.parent ).to.be.null;
+			expect( deserialized.name ).to.equal( 'el' );
+			expect( deserialized._children.length ).to.equal( 1 );
+
+			expect( deserialized._children._indexMap ).to.deep.equal( [ 0 ] );
```
Comment (pjasiun) [inline tests/treemodel/element.js]: Still, tests can not use (write nor read) private properties.

File: tests/treemodel/nodelist.js [hunk 29]
```diff
@@ -358,13 +359,125 @@ describe( 'NodeList', () => {
...
+			let nodeList = new NodeList( text );
+
+			let serialized = jsonParseStringify( nodeList );
+
+			let deserialized = NodeList.fromJSON( serialized );
+
+			expect( deserialized._indexMap ).to.deep.equal( nodeList._indexMap );
+			expect( deserialized._nodes.length ).to.equal( nodeList._nodes.length );
+			expect( deserialized._nodes[ 0 ].text ).to.equal( nodeList._nodes[ 0 ].text );
+			expect( [ ...deserialized._nodes[ 0 ]._attrs ] ).to.deep.equal( [ ...nodeList._nodes[ 0 ]._attrs ] );
```
Comment (pjasiun) [inline tests/treemodel/nodelist.js]: Same here. Tests can not use private properties. Use class API instead.

**Comment by pjasiun** [comment]
Much better :) This time only polishing (and private properties in tests) ;)

File: tests/treemodel/element.js (hunk 28 above)
Comment (jodator) [inline tests/treemodel/element.js]: But other test are using this :/ That's thy it is here. Just to note this, I'll change this in my changes.

File: src/treemodel/delta/deltafactory.js (hunk 26 above)
Comment (jodator) [inline src/treemodel/delta/deltafactory.js]: I think that is not possible with transpiled code (I'll check this again to be sure). There were some problems with that.

File: tests/treemodel/element.js (hunk 28 above)
Comment (pjasiun) [inline tests/treemodel/element.js]: Then please create a follow-up ticket. The Guilty will fix it ;)
